File: day15/solution.py
from pathlib import Path
from typing import List


CWD = Path(__file__).parent


def read_input() -> List[int]:
    input_file = CWD.joinpath("input.txt")
    with open(input_file, "r") as reader:
        return [int(n) for n in reader.readline().split(",")]


def pipe(n: int, numbers: List[int]) -> List[int]:
    res = numbers + [n]
    if len(res) > 2:
        res = res[1:]
    return res


def play(ins: List[int], max_turn: int) -> int:
    game = {n: index for index, n in enumerate(ins)}
    turn = len(ins) - 1
    last_said = ins[-1]
    while turn < max_turn-1:
        if last_said in game:
            v = turn - game[last_said]
        else:
            v = 0
        game[last_said] = turn
        last_said = v
        turn += 1
    return last_said


def first(ins: List[int]) -> int:
    return play(ins, 2020)


def second(ins: List[int]) -> int:
    return play(ins, 30000000)


def run() -> None:
    ins = read_input()
    print("First step:", first(ins)) # 257
    print("\nSecond step:", second(ins)) # 8546398

File: day15/solution.go
package day15

import (
	"io/ioutil"
	"log"
	"strconv"
	"strings"
	"time"
)

var input = "./input.txt"

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

// ReadInput retrieves the content of the input file
func ReadInput() (res []int) {
	data, _ := ioutil.ReadFile(input)
	for _, n := range strings.Split(string(data), ",") {
		val, _ := strconv.Atoi(n)
		res = append(res, val)
	}
	return
}

func play(numbers []int, maxTurn int) int {
	game := make(map[int]int)
	for index, n := range numbers {
		game[n] = index
	}
	lastSaid := numbers[len(numbers)-1]
	for turn := len(numbers) - 1; turn < maxTurn-1; turn++ {
		v := 0
		if n, ok := game[lastSaid]; ok {
			v = turn - n
		}
		game[lastSaid] = turn
		lastSaid = v
	}
	return lastSaid
}

// Step1 solves step 1
func Step1(numbers []int) int {
	defer timeTrack(time.Now(), "Step 1")
	return play(numbers, 2020)
}

// Step2 solves step 2
func Step2(numbers []int) int {
	defer timeTrack(time.Now(), "Step 2")
	return play(numbers, 30000000)
}
